shellhub-agent: Add service and package expressions

otavio opened the pull request:
###### Motivation for this change

We aim to ease the use of ShellHub and NixOS is one of the distros we'd like to support. It is working very well on our machines.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

otavio committed:
maintainer-list: add otavio

Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>

otavio committed:
nixos/shellhub-agent: initial service expression

This provides the service to configure the ShellHub Agent.

Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>

SuperSandro2000 commented on pkgs/applications/networking/shellhub-agent/default.nix:
```diff
@@ -0,0 +1,30 @@
+  version = "0.4.2";
+
+  src = fetchFromGitHub {
+    rev = "v${version}";
+    owner = "shellhub-io";
+    repo = "shellhub";
+    sha256 = "0cd41ing1pcf1bdaaq00w5h7lih5j2kcaa0m41g3ikm3vd1w5qna";
```
```suggestion
    owner = "shellhub-io";
    repo = "shellhub";
    rev = "v${version}";
    sha256 = "0cd41ing1pcf1bdaaq00w5h7lih5j2kcaa0m41g3ikm3vd1w5qna";
```

SuperSandro2000 commented on pkgs/applications/networking/shellhub-agent/default.nix:
```diff
@@ -0,0 +1,30 @@
+  buildFlagsArray = [ "-ldflags=-s -w -X main.AgentVersion=v${version}" ];
+
+  meta = with stdenv.lib; {
+    description =
```
The text is pretty long and fits better into longDescription. Can you add a shorter one which does not exceed about 80 characters?

otavio replied:
Sure; I'll rework it.

SuperSandro2000 commented on pkgs/applications/networking/shellhub-agent/default.nix:
```diff
@@ -0,0 +1,30 @@
+      "ShellHub is a modern SSH server for remotely accessing Linux devices via command line (using any SSH client) or web-based user interface, designed as an alternative to _sshd_. Think ShellHub as centralized SSH for the the edge and cloud computing.";
+    homepage = "https://shellhub.io/";
+    license = licenses.asl20;
+    maintainers = [ maintainers.otavio ];
```
```suggestion
    maintainers = with maintainers; [ otavio ];
```

SuperSandro2000 commented on pkgs/applications/networking/shellhub-agent/default.nix:
```diff
@@ -0,0 +1,30 @@
+    sha256 = "0cd41ing1pcf1bdaaq00w5h7lih5j2kcaa0m41g3ikm3vd1w5qna";
+  };
+
+  patches = [ ./go-mod-tidy.patch ];
```
It would be great if you could upstream those changes and then reference the patch from the commit hash in the PR. Like https://github.com/NixOS/nixpkgs/commit/51b455552402934e7a3abd49e843e46774078664.patch.

otavio replied:
We are working on the 0.5.0 release and I'll update it once it is released, dropping this patch.

SuperSandro2000 commented on nixos/modules/services/networking/shellhub-agent.nix:
```diff
@@ -0,0 +1,88 @@
+          Location where to store the ShellHub Agent private
+          key.
+        '';
+      };
+
+    };
+
+  };
```
```suggestion
      };
    };
  };
```

SuperSandro2000 commented on nixos/modules/services/networking/shellhub-agent.nix:
```diff
@@ -0,0 +1,88 @@
+      serviceConfig = {
+        Restart = "on-failure";
+        ExecStart = "${pkgs.shellhub-agent}/bin/agent";
+      };
+
+    };
```
```suggestion
      };
    };
```

SuperSandro2000 commented on nixos/modules/services/networking/shellhub-agent.nix:
```diff
@@ -0,0 +1,88 @@
+
+    };
+
+    environment.systemPackages = [ pkgs.shellhub-agent ];
+
+  };
+
+}
```
```suggestion
    environment.systemPackages = [ pkgs.shellhub-agent ];
  };
}
```

SuperSandro2000 commented on nixos/modules/services/networking/shellhub-agent.nix:
```diff
@@ -0,0 +1,88 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.shellhub-agent;
+
+in {
```
```suggestion
with lib;
let
  cfg = config.services.shellhub-agent;
in {
```

SuperSandro2000 commented on nixos/modules/services/networking/shellhub-agent.nix:
```diff
@@ -0,0 +1,88 @@
+
+      tenantId = mkOption {
+        type = types.str;
+        example = "ba0a880c-2ada-11eb-a35e-17266ef329d6";
```
What happens if someone uses this id? Would it just not work or do harm?

otavio replied:
The server provides this, so if you use an invalid tenant Id, it won't connect on the server-side.

SuperSandro2000 commented on nixos/modules/services/networking/shellhub-agent.nix:
```diff
@@ -0,0 +1,88 @@
+
+      privateKey = mkOption {
+        type = types.path;
+        default = "/var/lib/shellhub-agent/private.key";
```
I am not sure if this path works well with NixOS. It would be great if someone else could verify if this is okay.

otavio replied:
If there is a better way to do that, I am keen to learn and adapt.

otavio commented:
@SuperSandro2000 I succeed in clean it. Please do another review on this when you have time. Thanks for all the comments.

SuperSandro2000 commented on pkgs/applications/networking/shellhub-agent/default.nix:
```diff
@@ -0,0 +1,34 @@
+    homepage = "https://shellhub.io/";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ otavio ];
+    platforms = platforms.unix;
```
This is automatically set to ``linux ++ darwin`` in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/go-modules/generic/default.nix#L239. This should be removed here and only be set if only linux or darwin is supported.

otavio replied:
Oh, good point. I fixed it and pushed it again.

SuperSandro2000 reviewed (commented):
I am not that familiar with modules so someone else should take a look at it.

otavio commented:
Is there anything else pending for this PR?

SuperSandro2000 commented on maintainers/maintainer-list.nix:
```diff
@@ -6563,6 +6563,12 @@
     githubId = 111265;
     name = "Ozan Sener";
   };
+  otavio = {
```
Please put this into a separate commit with the message ``maintainers: add otavio``.

otavio replied:
This is already in a specific commit; I reworked the commit log as you suggested.

SuperSandro2000 commented on pkgs/applications/networking/shellhub-agent/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+  meta = with stdenv.lib; {
+    description =
+      "ShellHub enables easy access any Linux device behind firewall and NAT";
```
```suggestion
      "Enables easy access any Linux device behind firewall and NAT";
```

otavio replied:
Applied.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 104419` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

SuperSandro2000 commented:
Result of `nixpkgs-review pr 104419` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>shellhub-agent</li>
  </ul>
</details>

otavio commented:
Hello all, I rebased this on top of master to resolve the conflict. Could someone take a look at this?

jonringer commented on pkgs/applications/networking/shellhub-agent/default.nix:
```diff
@@ -0,0 +1,34 @@
+{ stdenv, buildGoModule, fetchFromGitHub }:
+
+buildGoModule rec {
+  pname = "shellhub-agent";
+  version = "0.4.2";
+
```
This should probably be a different commit:
```
shellhub-agent: init at 0.4.2
```

otavio replied:
Done.

jonringer reviewed (commented):
I would break out the package and service into 2 different ocmmits:
```
shellhub-agent: init at 0.4.2
nixos/shellhub-agent: init
```

otavio committed:
shellhub-agent: initial expression at 0.4.2

ShellHub is a modern SSH server for remotely accessing Linux devices via
command line (using any SSH client) or web-based user interface,
designed as an alternative to _sshd_.

Think ShellHub as centralized SSH for the the edge and cloud computing.

More information about the project in:

 - https://shellhub.io
 - https://github.com/shellhub-io/shellhub

Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>

otavio commented on nixos/modules/services/networking/shellhub-agent.nix:
```diff
@@ -0,0 +1,80 @@
+      };
+    };
+
+    environment.systemPackages = [ pkgs.shellhub-agent ];
```
Done.

aanderse replied:
Or overlays. Don't forget those!

otavio replied:
Absolutely. I reworked the commits as suggested by @jonringer, so I think it is fine now. @aanderse, please take a look as well.

jonringer reviewed (approved):
LGTM

cc @Mic92  or @Infinisil for final look-over of module

Infinisil commented on nixos/modules/services/networking/shellhub-agent.nix:
```diff
@@ -0,0 +1,89 @@
+
+      serviceConfig = {
+        Restart = "on-failure";
+        ExecStart = "${cfg.package}/bin/agent";
```
This runs the service as root. Is this required? If it is, then I suggest explicitly setting `User = "root"` and adding a comment why it's necessary. It may also be possible to not require root, but just some special capabilities, in which case you could try `DynamicUser = true` and `AmbientCapabalities`/`CapabilityBoundingSet` (see other modules for examples).

otavio replied:
Thanks @Infinisil for the review. For now, we need to use root but I'll investigate if in the future we can drop the capabilities.

Infinisil replied:
You added a comment mentioning that it requires root, but it doesn't explain _why_ that's the case, which is what I'm looking for (probably just because it needs to log into arbitrary users?)

otavio replied:
It is an SSH server so yes, it needs to initialize sessions with different users. I assumed this was implicit by it being an SSH server. Should I make this explicit there?

otavio replied:
I added this explicitly there.

infinisil reviewed (approved):
That's great, thanks :)
